src/log_surgeon/finite_automata/DeterminizationConfiguration.hpp

```diff
@@ -0,0 +1,157 @@
+#ifndef LOG_SURGEON_FINITE_AUTOMATA_DETERMINIZATION_CONFIGURATION_HPP
+#define LOG_SURGEON_FINITE_AUTOMATA_DETERMINIZATION_CONFIGURATION_HPP
+
+#include <map>
+#include <optional>
+#include <set>
+#include <stack>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+#include <log_surgeon/finite_automata/TagOperation.hpp>
+#include <log_surgeon/types.hpp>
+
+namespace log_surgeon::finite_automata {
+/**
+ * Encapsulates the configuration of a state during the determinization process of an NFA.
+ *
+ * This class stores the NFA state, the mapping of tag IDs to register IDs, the history of tag
+ * operations, and the lookahead of upcoming tag operations.
+ *
+ * The configuration also supports exploring reachable configurations via spontaneous transitions.
+ *
+ * @tparam TypedNfaState The type of the NFA state.
+ */
+template <typename TypedNfaState>
+class DeterminizationConfiguration {
+public:
+    DeterminizationConfiguration(
+            TypedNfaState const* nfa_state,
+            std::map<tag_id_t, reg_id_t> tag_to_reg_ids,
+            std::vector<TagOperation> tag_history,
+            std::vector<TagOperation> tag_lookahead
+    )
+            : m_nfa_state{nfa_state},
+              m_tag_id_to_reg_ids{std::move(tag_to_reg_ids)},
+              m_history{std::move(tag_history)},
+              m_lookahead{std::move(tag_lookahead)} {
```

Comment:
```suggestion
            std::map<tag_id_t, reg_id_t> tag_id_to_reg_ids,
            std::vector<TagOperation> tag_history,
            std::vector<TagOperation> tag_lookahead
    )
            : m_nfa_state{nfa_state},
              m_tag_id_to_reg_ids{std::move(tag_id_to_reg_ids)},
              m_history{std::move(tag_history)},
              m_lookahead{std::move(tag_lookahead)} {
```
Trying to make things consistent since we tend to use `tag_id` in mappings.